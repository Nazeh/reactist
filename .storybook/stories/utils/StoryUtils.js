import React from 'react'
import { withInfo } from '@storybook/addon-info'

// Info Addon =================================================================
const getPropTypesStory = (...components) => withInfo({
    header: false,
    inline: true,
    source: false,
    propTables: components,
    maxPropStringLength: 300
})(() => <div />)

// Chapters Addon =============================================================
const optionsSourceOnly = {
    showSource: false,
    allowSourceToggling: true,
    showPropTables: false,
    allowPropTablesToggling: false
}
const optionsNoSourceNoProps = {
    showSource: false,
    allowSourceToggling: false,
    showPropTables: false,
    allowPropTablesToggling: false
}
const emptySection = () => {}

export {
    getPropTypesStory,
    emptySection,
    optionsSourceOnly,
    optionsNoSourceNoProps
}
