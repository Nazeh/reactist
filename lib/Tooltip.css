.reactist.tooltip {
  background-color: #333333;
  border-radius: 3px;
  padding: 5px 10px;
  max-width: 250px;
  white-space: normal;
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  visibility: hidden;
  opacity: 0;
  transition: opacity 0.1s linear;
}
.reactist.tooltip__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist.tooltip .tooltip__text {
  font-size: 0.81rem;
  color: white;
  font-weight: 400;
  line-height: 1.6;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  display: inline-block;
}
.reactist.tooltip.visible {
  visibility: visible;
  opacity: 1;
}
.reactist.tooltip.inverted {
  background-color: white;
  border-radius: 3px;
  border: 1px solid #dcdcdc;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
}
.reactist.tooltip.inverted .tooltip__text {
  color: #202020;
}
.reactist.tooltip.arrow_top:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #333333;
}
.reactist.tooltip.arrow_top.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #dcdcdc;
  top: calc(-5px - 1px);
}
.reactist.tooltip.arrow_top.inverted:after {
  border-bottom: 5px solid white;
}
.reactist.tooltip.arrow_right:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #333333;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist.tooltip.arrow_right.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
  right: calc(-5px - 1px);
}
.reactist.tooltip.arrow_right.inverted:after {
  border-left: 5px solid white;
}
.reactist.tooltip.arrow_bottom:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #333333;
}
.reactist.tooltip.arrow_bottom.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #dcdcdc;
  bottom: calc(-5px - 1px);
}
.reactist.tooltip.arrow_bottom.inverted:after {
  border-top: 5px solid white;
}
.reactist.tooltip.arrow_left:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #333333;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist.tooltip.arrow_left.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
  left: calc(-5px - 1px);
}
.reactist.tooltip.arrow_left.inverted:after {
  border-right: 5px solid white;
}


/*# sourceMappingURL=Tooltip.map*/