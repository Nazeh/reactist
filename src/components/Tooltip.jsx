import './styles/tooltip.less'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import {
    hasEnoughSpace,
    calculatePosition
} from './utils/PositioningUtils'

class Tooltip extends React.Component {
    constructor(props, context) {
        super(props, context)
        this.state = { visible: false }
    }

    shouldComponentUpdate(nextProps, nextState) {
        // only update on state or prop changes
        return this.state.visible !== nextState.visible ||
            this.props.position !== nextProps.position ||
            this.props.text !== nextProps.text ||
            this.props.hideOnScroll !== nextProps.hideOnScroll ||
            this.props.delayShow !== nextProps.delayShow ||
            this.props.delayHide !== nextProps.delayHide ||
            this.props.children !== nextProps.children
    }

    componentDidUpdate() {
        if (this.wrapper && this.state.visible) {
            // Interact with the DOM after a state update
            this._updateTooltipPosition()
        }
    }

    componentWillUnmount() {
        this._clearDelayTimeout()
        this._removeScrollListener()
    }

    _initScrollListener() {
        document.addEventListener('scroll', this._hide)
    }

    _removeScrollListener() {
        document.removeEventListener('scroll', this._hide)
    }

    _clearDelayTimeout() {
        if (this.delayTimeout) {
            clearTimeout(this.delayTimeout)
        }
    }

    _show = () => {
        this._delayAction(() => {
            this.setState(() => ({ visible: true }))
            if (this.props.hideOnScroll) {
                this._initScrollListener()
            }
        }, this.props.delayShow)
    }
    _hide = () => {
        this._delayAction(() => {
            this._clearDelayTimeout()
            this.setState(() => ({ visible: false }))
            if (this.props.hideOnScroll) {
                this._removeScrollListener()
            }
        }, this.props.delayHide)
    }

    _delayAction(actionFn, delay) {
        this._clearDelayTimeout()
        this.delayTimeout = setTimeout(actionFn, delay)
    }

    _updateTooltipPosition = () => {
        const { position, allowVaguePositioning } = this.props
        const gapSize = 5 // size of the arrow (see `tooltip.less`)
        const wrapperRect = this.wrapper.getBoundingClientRect()
        const tooltipRect = this.tooltip.getBoundingClientRect()

        const windowDimensions = {
            height: Math.max(document.documentElement.clientHeight, window.innerHeight),
            width: Math.max(document.documentElement.clientWidth, window.innerWidth)
        }
        const tooltipDimensions = {
            height: tooltipRect.height,
            width: tooltipRect.width
        }
        const wrapperDimensions = {
            height: wrapperRect.height,
            width: wrapperRect.width
        }
        const wrapperPosition = {
            x: wrapperRect.x,
            y: wrapperRect.y
        }

        const positionsToTry = position === 'auto'
            ? ['top', 'right', 'bottom', 'left', 'top']
            : [position]

        for (let index = 0; index < positionsToTry.length; index++) {
            const currentPosition = positionsToTry[index]
            const enoughSpaceAtPosition = hasEnoughSpace(windowDimensions, tooltipDimensions, wrapperDimensions, wrapperPosition, currentPosition, gapSize)

            if (enoughSpaceAtPosition || index === positionsToTry.length - 1) {
                const tooltipPosition = calculatePosition(currentPosition, wrapperDimensions, wrapperPosition, tooltipDimensions, gapSize)
                this.tooltip.style.top = `${tooltipPosition.y}px`
                this.tooltip.style.left = tooltipPosition.x < 0 && allowVaguePositioning
                    ? `${2 * gapSize}px` // not centered but fully visible
                    : `${tooltipPosition.x}px`

                if (currentPosition !== position) {
                    this.tooltip.className = this._getClassNameForPosition(currentPosition)
                }
                break
            }
        }
    }

    _getClassNameForPosition = (position) => {
        const { visible } = this.state
        const { tooltipClassName } = this.props
        const className = classNames('reactist tooltip', { visible }, tooltipClassName)
        if (visible) {
            return classNames(className, {
                arrow_top: position === 'bottom',
                arrow_right: position === 'left',
                arrow_bottom: position === 'auto' || position === 'top',
                arrow_left: position === 'right'
            })
        }
        return className
    }

    render() {
        if (!this.props.text) {
            return (
                <div className='reactist tooltip__wrapper'>
                    { this.props.children }
                </div>
            )
        }

        const tooltipClass = this._getClassNameForPosition(this.props.position)
        const wrapperClass = classNames('reactist tooltip__wrapper', this.props.wrapperClassName)
        return (
            <span
                className={ wrapperClass }
                onMouseEnter={ this._show }
                onMouseLeave={ this._hide }
                ref={ wrapper => this.wrapper = wrapper }
            >
                { this.props.children }
                <span className={ tooltipClass } ref={ tooltip => this.tooltip = tooltip }>
                    <span className='tooltip__text'>{ this.props.text }</span>
                </span>
            </span>
        )
    }
}
Tooltip.displayName = 'Tooltip'
Tooltip.defaultProps = {
    position: 'auto',
    hideOnScroll: true,
    delayShow: 1000,
    delayHide: 0,
    allowVaguePositioning: false
}
Tooltip.propTypes = {
    /**
     * Position of the tooltip. Defaults to `auto`.
     * `auto` tries to position the tooltip to the top,
     * if there's not enough space it tries to position the tooltip clockwise (right, bottom, left).
     * Setting a distinct value like `right` will always position the tooltip right, regardless of available space.
     */
    position: PropTypes.oneOf(['auto', 'top', 'right', 'bottom', 'left']),
    /**
     * Whether vague positioning is allowed. When set to true the tooltip prefers to be fully visible over being correctly centered.
     */
    allowVaguePositioning: PropTypes.bool,
    /** Text that is displayed inside the tooltip */
    text: PropTypes.string.isRequired,
    /** Delay before the tooltip appears and disappears (in ms). */
    /** Set whether scrolling should hide the tooltip or not. */
    hideOnScroll: PropTypes.bool,
    /** How long to wait after hovering before the tooltip is shown (in ms). */
    delayShow: PropTypes.number,
    /** How long to wait after unhovering before the tooltip is hidden (in ms). */
    delayHide: PropTypes.number,
    /** Children that are wrapped by the toolip. */
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]),
    /** Additional css class that is applied to the wrapper element. */
    wrapperClassName: PropTypes.string,
    /** Additional css class that is applied to the tooltip element. */
    tooltipClassName: PropTypes.string
}

export default Tooltip
