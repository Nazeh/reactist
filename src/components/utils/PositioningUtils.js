const hasEnoughSpace = (windowDimensions, elementDimensions, wrapperDimensions, wrapperPosition, position, gap = 0) => {
    const { height: windowHeight, width: windowWidth } = windowDimensions
    const { height: elementHeight, width: elementWidth } = elementDimensions
    const { height: wrapperHeight, width: wrapperWidth } = wrapperDimensions
    const { x: wrapperX, y: wrapperY } = wrapperPosition

    const verticalPosition = _calculateVerticalPosition(wrapperPosition, wrapperDimensions, elementDimensions)
    const horizontalPosition = _calculateHorizontalPosition(wrapperPosition, wrapperDimensions, elementDimensions)
    const canPlaceVertically = verticalPosition + elementWidth <= windowWidth
    const canPlaceHorizontally = horizontalPosition - elementHeight >= 0

    if (position === 'top') {
        return canPlaceVertically
            && (wrapperY - elementHeight - gap) >= 0
    } else if (position === 'right') {
        return canPlaceHorizontally
            && (wrapperX + wrapperWidth + elementWidth + gap) <= windowWidth
    } else if (position === 'left') {
        return canPlaceHorizontally
            && (wrapperX - elementWidth - gap) >= 0
    } else if (position === 'bottom') {
        return canPlaceVertically
            && (wrapperY + wrapperHeight + elementHeight + gap) <= windowHeight
    }
    return false
}

const _calculateVerticalPosition = (wrapperPosition, wrapperDimensions, elementDimensions) => {
    return wrapperPosition.x + ((wrapperDimensions.width - elementDimensions.width) / 2)
}
const _calculateHorizontalPosition = (wrapperPosition, wrapperDimensions, elementDimensions) => {
    return wrapperPosition.y + ((wrapperDimensions.height - elementDimensions.height) / 2)
}

const calculateTopCenterPosition = (wrapperDimensions, wrapperPosition, elementDimensions, gap = 0) => {
    const x = _calculateVerticalPosition(wrapperPosition, wrapperDimensions, elementDimensions)
    const y = wrapperPosition.y - elementDimensions.height - gap
    return { x, y }
}

const calculateBottomCenterPosition = (wrapperDimensions, wrapperPosition, elementDimensions, gap = 0) => {
    const x = _calculateVerticalPosition(wrapperPosition, wrapperDimensions, elementDimensions)
    const y = wrapperPosition.y + wrapperDimensions.height + gap
    return { x, y }
}

const calculateRightCenterPosition = (wrapperDimensions, wrapperPosition, elementDimensions, gap = 0) => {
    const x = wrapperPosition.x + wrapperDimensions.width + gap
    const y = _calculateHorizontalPosition(wrapperPosition, wrapperDimensions, elementDimensions)
    return { x, y }
}

const calculateLeftCenterPosition = (wrapperDimensions, wrapperPosition, elementDimensions, gap = 0) => {
    const x = wrapperPosition.x - elementDimensions.width - gap
    const y = _calculateHorizontalPosition(wrapperPosition, wrapperDimensions, elementDimensions)
    return { x, y }
}

const calculatePosition = (position, wrapperDimensions, wrapperPosition, elementDimensions, gap = 0)  => {
    if (position === 'top') {
        return calculateTopCenterPosition(wrapperDimensions, wrapperPosition, elementDimensions, gap)
    } else if (position === 'right') {
        return calculateRightCenterPosition(wrapperDimensions, wrapperPosition, elementDimensions, gap)
    } else if (position === 'bottom') {
        return calculateBottomCenterPosition(wrapperDimensions, wrapperPosition, elementDimensions, gap)
    } else if (position === 'left') {
        return calculateLeftCenterPosition(wrapperDimensions, wrapperPosition, elementDimensions, gap)
    }
    return wrapperPosition
}

export {
    hasEnoughSpace,
    calculatePosition,
    calculateTopCenterPosition,
    calculateBottomCenterPosition,
    calculateRightCenterPosition,
    calculateLeftCenterPosition
}
