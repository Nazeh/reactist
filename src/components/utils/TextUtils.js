const getInitials = name => {
    if (!name) {
        return ''
    }

    const seed = name.trim().split(' ')
    const first_initial = seed[0]
    const last_initial = seed[seed.length - 1]

    let initials = first_initial[0]
    if (first_initial[0] !== last_initial[0]) {
        initials += last_initial[0]
    }
    return initials.toUpperCase()
}

const emailToIndex = (email, maxIndex) => {
    const seed = email.split('@')[0]
    const hash = seed.charCodeAt(0) + seed.charCodeAt(seed.length - 1) || 0
    return hash % maxIndex
}

export { getInitials, emailToIndex }
