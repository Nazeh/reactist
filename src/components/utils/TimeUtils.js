import moment from 'moment'

const TimeUtils = {
    SHORT_FORMAT_CURRENT_YEAR: 'MMM D',
    SHORT_FORMAT_PAST_YEAR: 'MMM D, YYYY',
    LONG_FORMAT: 'DD MMM YY, LT',

    timeAgo(timestamp, config = {}) {
        const {
            locale = 'en',
            hoursSuffix = 'h',
            minutesSuffix = 'm',
            momentsAgo = 'moments ago'
        } = config
        const now = moment()
        const date = moment.unix(timestamp)
        date.locale(locale)
        const diffMinutes = now.diff(date, 'minutes')
        const diffHours = now.diff(date, 'hours')
        const diffDays = now.diff(date, 'days')

        if (diffDays > 7) {
            if (date.isSame(now, 'year')) {
                return date.format(this.SHORT_FORMAT_CURRENT_YEAR)
            } else {
                return date.format(this.SHORT_FORMAT_PAST_YEAR)
            }
        } else if (diffDays > 0 && diffDays <= 7) {
            return date.format('dddd')
        } else if (diffHours > 0 && diffHours <= 23) {
            return `${diffHours}${hoursSuffix}`
        } else if (diffMinutes > 0 && diffMinutes <= 59) {
            return `${diffMinutes}${minutesSuffix}`
        } else {
            return momentsAgo
        }
    },

    formatTime(timestamp, locale = 'en') {
        const date = moment.unix(timestamp)
        date.locale(locale)
        if (date.isSame(moment(), 'year')) {
            return date.format(this.SHORT_FORMAT_CURRENT_YEAR)
        } else {
            return date.format(this.SHORT_FORMAT_PAST_YEAR)
        }
    },

    formatTimeLong(timestamp, locale = 'en') {
        const date = moment.unix(timestamp)
        date.locale(locale)
        return date.format(this.LONG_FORMAT)
    }
}

export default TimeUtils
