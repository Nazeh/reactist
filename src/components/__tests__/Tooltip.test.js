import React from 'react'
import { shallow, mount } from 'enzyme'

import Tooltip from '../Tooltip'
import * as PositioningUtils from '../utils/PositioningUtils'

describe('Tooltip', () => {
    // Setup ==================================================================
    beforeEach(() => {
        jest.useFakeTimers()
    })
    afterEach(() => {
        jest.useRealTimers()
    })

    it('renders without crashing', () => {
        const tooltip = shallow(getTooltip())
        expect(tooltip).toMatchSnapshot()
    })

    it('renders only children if text prop is not defined', () => {
        // eslint-disable-next-line
        console.error = jest.fn() // silence errors caused by unsupplied required prop
        const tooltip = shallow(getTooltip({ text: null }))
        expect(tooltip).toMatchSnapshot()
    })

    it('gets visible on hovering and hides on unhovering', () => {
        const tooltip = shallow(getTooltip())

        tooltip.simulate('mouseEnter')
        jest.runAllTimers()
        expect(tooltip.state().visible).toBe(true)

        tooltip.simulate('mouseLeave')
        jest.runAllTimers()
        expect(tooltip.state().visible).toBe(false)
    })

    describe('Arrow Placement', () => {
        it('draws a bottom arrow when rendering to the top', () => {
            const tooltip = shallow(getTooltip({ position: 'top' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_bottom')).toBeTruthy()
        })

        it('draws a left arrow when rendering to the right', () => {
            const tooltip = shallow(getTooltip({ position: 'right' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_left')).toBeTruthy()
        })

        it('draws a top arrow when rendering to the bottom', () => {
            const tooltip = shallow(getTooltip({ position: 'bottom' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_top')).toBeTruthy()
        })

        it('draws a right arrow when rendering to the left', () => {
            const tooltip = shallow(getTooltip({ position: 'left' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_right')).toBeTruthy()
        })
    })

    describe('Scroll Listener', () => {
        it('adds a capturing scrolllistener when becoming visible and when hideOnScroll is true', () => {
            document.addEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: true }))

            tooltip.simulate('mouseEnter')
            jest.runAllTimers()
            const callArgs = document.addEventListener.mock.calls[0]
            expect(callArgs[0]).toBe('scroll')
            expect(callArgs[2]).toBe(true)
        })

        it('does not add a scrolllistener when becoming visible and when hideOnScroll is false', () => {
            document.addEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: false }))

            tooltip.simulate('mouseEnter')
            jest.runAllTimers()
            expect(document.addEventListener).not.toHaveBeenCalled()
        })

        it('removes a capturing scrolllistener when hiding and when hideOnScroll is true', () => {
            document.removeEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: true }))

            tooltip.simulate('mouseLeave')
            jest.runAllTimers()
            const callArgs = document.removeEventListener.mock.calls[0]
            expect(callArgs[0]).toBe('scroll')
            expect(callArgs[2]).toBe(true)
        })

        it('does not remove a scrolllistener when hiding and when hideOnScroll is false', () => {
            document.removeEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: false }))

            tooltip.simulate('mouseLeave')
            jest.runAllTimers()
            expect(document.removeEventListener).not.toHaveBeenCalled()
        })

        it('removes scroll listener on unmount', () => {
            document.removeEventListener = jest.fn()
            const tooltip = shallow(getTooltip()).instance()

            tooltip.componentWillUnmount()
            expect(document.removeEventListener).toHaveBeenCalled()
        })
    })

    describe('Position Calculation', () => {
        it('calculates tooltip position after becoming visible', () => {
            const updateTooltipPositionSpy = jest.fn()
            const tooltip = mount(getTooltip()).instance()
            tooltip._updateTooltipPosition = updateTooltipPositionSpy

            tooltip.setState({ visible: true })
            expect(updateTooltipPositionSpy).toHaveBeenCalled()
        })

        it('sets the tooltip to the given position if it is not set to `auto`', () => {
            PositioningUtils.hasEnoughSpace = jest.fn(() => true)
            jest.spyOn(
                PositioningUtils,
                'calculatePosition'
            ).mockImplementationOnce(() => ({
                x: 23,
                y: 42
            }))

            const tooltip = mount(getTooltip({ position: 'right' })).instance()
            tooltip.setState({ visible: true }) // triggers update positions

            expect(tooltip.tooltip.style.getPropertyValue('top')).toBe('42px')
            expect(tooltip.tooltip.style.getPropertyValue('left')).toBe('23px')
        })

        it('allows vague positioning to avoid cut offs', () => {
            PositioningUtils.hasEnoughSpace = jest.fn(() => true)
            jest.spyOn(
                PositioningUtils,
                'calculatePosition'
            ).mockImplementationOnce(() => ({
                x: -23,
                y: 42
            }))

            const tooltip = mount(
                getTooltip({ allowVaguePositioning: true })
            ).instance()
            tooltip.setState({ visible: true }) // triggers update positions

            expect(tooltip.tooltip.style.getPropertyValue('top')).toBe('42px')
            expect(tooltip.tooltip.style.getPropertyValue('left')).toBe('10px')
        })

        it('sets the tooltip to the first position that has enough space when `auto` is supplied', () => {
            PositioningUtils.hasEnoughSpace = jest
                .fn()
                .mockReturnValueOnce(false) // top
                .mockReturnValueOnce(false) // right
                .mockReturnValueOnce(false) // bottom
                .mockReturnValueOnce(true) // left
            jest.spyOn(
                PositioningUtils,
                'calculatePosition'
            ).mockImplementationOnce(() => ({
                x: 23,
                y: 42
            }))
            const tooltip = mount(getTooltip({ position: 'auto' })).instance()
            tooltip.setState({ visible: true }) // triggers update positions

            expect(tooltip.tooltip.style.getPropertyValue('top')).toBe('42px')
            expect(tooltip.tooltip.style.getPropertyValue('left')).toBe('23px')
            expect(tooltip.tooltip.className).toContain('arrow_right')
        })

        it('sets the tooltip to the correct position when changing the gap size', () => {
            PositioningUtils.hasEnoughSpace = jest.fn(() => true)

            const wrapper = mount(getTooltip({ position: 'top' }))
            const instance = wrapper.instance()

            instance.wrapper.getBoundingClientRect = jest.fn(() => ({
                left: 500,
                top: 500,
                width: 70,
                height: 20
            }))

            instance.tooltip.getBoundingClientRect = jest.fn(() => ({
                width: 70,
                height: 20
            }))

            instance.setState({ visible: true }) // triggers update positions

            expect(instance.tooltip.style.getPropertyValue('top')).toBe('475px')

            wrapper.setProps({ gapSize: 20 })

            expect(instance.tooltip.style.getPropertyValue('top')).toBe('460px')
        })
    })

    describe('Component Updates', () => {
        it('updates when the state changes', () => {
            const tooltip = shallow(getTooltip()).instance()

            const currentProps = tooltip.props
            const shouldUpdate = tooltip.shouldComponentUpdate(currentProps, {
                visible: true
            })
            expect(shouldUpdate).toBe(true)
        })

        it('updates when any property changes', () => {
            const tooltip = shallow(getTooltip()).instance()
            const currentState = tooltip.state
            const currentProps = tooltip.props

            const shouldUpdateAfterPositionChange = tooltip.shouldComponentUpdate(
                { ...currentProps, position: 'bottom' },
                currentState
            )
            expect(shouldUpdateAfterPositionChange).toBe(true)

            const shouldUpdateAfterTextChange = tooltip.shouldComponentUpdate(
                { ...currentProps, text: 'new tip' },
                currentState
            )
            expect(shouldUpdateAfterTextChange).toBe(true)

            const shouldUpdateAfterHideOnScrollChange = tooltip.shouldComponentUpdate(
                { ...currentProps, hideOnScroll: false },
                currentState
            )
            expect(shouldUpdateAfterHideOnScrollChange).toBe(true)

            const shouldUpdateAfterDelayShowChange = tooltip.shouldComponentUpdate(
                { ...currentProps, delayShow: 2342 },
                currentState
            )
            expect(shouldUpdateAfterDelayShowChange).toBe(true)

            const shouldUpdateAfterDelayHideChange = tooltip.shouldComponentUpdate(
                { ...currentProps, delayHide: 2342 },
                currentState
            )
            expect(shouldUpdateAfterDelayHideChange).toBe(true)

            const shouldUpdateAfterGapSizeChange = tooltip.shouldComponentUpdate(
                { ...currentProps, gapSize: 10 },
                currentState
            )
            expect(shouldUpdateAfterGapSizeChange).toBe(true)

            const shouldUpdateAfterChildrenChange = tooltip.shouldComponentUpdate(
                { ...currentProps, children: <span>New Content</span> },
                currentState
            )
            expect(shouldUpdateAfterChildrenChange).toBe(true)
        })

        it('does not update when state and properties did not change', () => {
            const tooltip = shallow(getTooltip()).instance()

            const currentState = tooltip.state
            const currentProps = tooltip.props
            const shouldUpdate = tooltip.shouldComponentUpdate(
                currentProps,
                currentState
            )
            expect(shouldUpdate).toBe(false)
        })
    })

    it('sets ref of tooltip and wrapper after mounting', () => {
        const tooltip = mount(getTooltip()).instance()

        expect(tooltip.wrapper).toBeTruthy()
        expect(tooltip.tooltip).toBeTruthy()
    })

    // Helpers ================================================================
    const getTooltip = (props = {}) => (
        <Tooltip text="tip" {...props}>
            <span>Wrapped Content</span>
        </Tooltip>
    )
})
