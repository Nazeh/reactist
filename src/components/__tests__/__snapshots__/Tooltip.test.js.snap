// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders only children if text prop is not defined 1`] = `
<div
  className="reactist tooltip__wrapper"
>
  <span>
    Wrapped Content
  </span>
</div>
`;

exports[`Tooltip renders without crashing 1`] = `
<span
  className="reactist tooltip__wrapper"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span>
    Wrapped Content
  </span>
  <span
    className="reactist tooltip"
  >
    <span
      className="tooltip__text"
    >
      tip
    </span>
  </span>
</span>
`;
